src/compiler.jl

```diff
@@ -809,93 +1117,283 @@ function runtime_invoke_fwd(arg_ptr, shadow_ptr, activity_ptr::Val{ActivityTup},
+end
+
+const FwdCache = Dict{Tuple{Int64,Int64},Function}()
+function runtime_generic_fwd(N::Int64,Width::Int64,create::Bool=false)::Function
+    if haskey(FwdCache, (N,Width))
+        return FwdCache[(N,Width)]
+    end
+    if !create
+        return runtime_generic_fwd_fallback
+    end
+    primargs, shadowargs, primtypes, allargs, typeargs, wrapped = setup_macro_wraps(true, N, Width)
+    funcname = Symbol("inner_runtime_generic_fwd_"*string(N)*"_"*string(Width))
+    efn = quote
+        function $funcname(activity::Val{ActivityTup}, width::Val{$Width}, RT::Val{ReturnType}, f::F, df::DF,$(allargs...)) where {ActivityTup,ReturnType, F, DF, $(typeargs...)}
+            args = ($(wrapped...),)
+
+            fn = f
+            dfn = ActivityTup[1] ? df : nothing
+
+            # TODO: Annotation of return value
+            # tt0 = Tuple{$(primtypes...)}
+            tt = Tuple{map(x->eltype(typeof(x)), args)...}
+            tt′ = Tuple{map(typeof, args)...}
+            rt = Core.Compiler.return_type(fn, tt)
+            annotation = guess_activity(rt, API.DEM_ForwardMode)
+            
+            if annotation <: DuplicatedNoNeed
+                annotation = Duplicated{rt}
+            end
+            if $Width != 1
+                if annotation <: Duplicated
+                    annotation = BatchDuplicated{rt, $Width}
+                end
+            end
 
-    if length(res) == 0
-        return ReturnType(ntuple(i->nothing, Val(Width+1)))
+            forward = thunk(fn, dfn, annotation, tt′, Val(API.DEM_ForwardMode), width, #=ModifiedBetween=#Val(false), #=returnPrimal=#Val(true))
+
+            res = forward(args...)
+
+            if length(res) == 0
+                return ReturnType(ntuple(i->nothing, Val($Width+1)))
+            end
+            if annotation <: Const
+                return ReturnType(ntuple(i->res[1], Val($Width+1)))
+            end
+
+            if $Width == 1
+               return ReturnType((res[1], res[2]))
+            else
+                return ReturnType((res[1], res[2]...))
+            end
+        end
     end
-    if annotation <: Const
-        return ReturnType(ntuple(i->res[1], Val(Width+1)))
+    fn = eval(efn)
+    FwdCache[(N,Width)] = fn
+    return fn
+
+end
+
+const AugCache = Dict{Tuple{Int64,Int64},Function}()
+function runtime_generic_augfwd(N::Int64,Width::Int64,create::Bool=false)::Function
+    if haskey(AugCache, (N,Width))
+        return AugCache[(N,Width)]
+    end
+    if !create
+        return runtime_generic_augfwd_fallback
     end
+    primargs, shadowargs, primtypes, allargs, typeargs, wrapped = setup_macro_wraps(false, N, Width)
+    funcname = Symbol("inner_runtime_generic_augfwd_"*string(N)*"_"*string(Width))
+    fn = eval(quote
+        function $funcname(activity::Val{ActivityTup}, width::Val{$Width}, RT::Val{ReturnType}, f::F, df::DF,$(allargs...)) where {ActivityTup,ReturnType, F, DF, $(typeargs...)}
```

Comment:
Why do we need to specialize on Width and allargs?

Can we not specialize on first call instead of going through this whole custom cache nonsense?

Reply:
Having varargs prevents the correct type from being deduced, hence the splatted type. The caching is necessary because otherwise eval'ing a function isn't legal from a global scope. I would vastly prefer getting rid of the cache and specializing on the first call if you have ideas.